controllers/managedfusiondeployment_controller.go

```diff
@@ -0,0 +1,750 @@
+			},
+		),
+	)
+	enqueueManagedFusionAgentRequest := handler.EnqueueRequestsFromMapFunc(
+		func(object client.Object) []reconcile.Request {
+			return []reconcile.Request{{
+				NamespacedName: types.NamespacedName{
+					Name:      managedFusionSecretName,
+					Namespace: r.Namespace,
+				},
+			}}
+		},
+	)
+	return ctrl.NewControllerManagedBy(mgr).
+		WithOptions(*ctrlOptions).
+		For(&corev1.Secret{}, managedFusionSecretPredicates).
+		// Watch owned resources
+		Owns(&promv1.Prometheus{}).
+		Owns(&promv1.Alertmanager{}).
+		Owns(&corev1.Secret{}).
+		Owns(&promv1a1.AlertmanagerConfig{}).
+		Owns(&corev1.ConfigMap{}).
+		Owns(&corev1.Service{}).
+		Owns(&promv1.ServiceMonitor{}).
+		Owns(&netv1.NetworkPolicy{}).
+		Watches(
+			&source.Kind{Type: &promv1.PodMonitor{}},
+			enqueueManagedFusionAgentRequest,
+			monResourcesPredicates,
+		).
+		Watches(
+			&source.Kind{Type: &promv1.ServiceMonitor{}},
+			enqueueManagedFusionAgentRequest,
+			monResourcesPredicates,
+		).
+		Watches(
+			&source.Kind{Type: &promv1.PrometheusRule{}},
+			enqueueManagedFusionAgentRequest,
+			monResourcesPredicates,
+		).
+		Complete(r)
+}
+
+// Reconcile changes to all owned resource based on the infromation provided by the ManangedFusionDeployment secret resource
+func (r *ManagedFusionReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := r.Log.WithValues("req.Namespace", req.Namespace, "req.Name", req.Name)
+	log.Info("Starting reconcile for ManangedFusionDeployment")
+
+	r.initReconciler(ctx, req)
+	if err := r.get(r.managedFusionSecret); err != nil {
+		if errors.IsNotFound(err) {
+			r.Log.V(-1).Info(fmt.Sprintf("%s secret resource not found", managedFusionSecretName))
+			return ctrl.Result{}, nil
+		}
+		r.Log.Error(err, fmt.Sprintf("Failed to get %s secret", managedFusionSecretName))
+		return ctrl.Result{}, fmt.Errorf("failed to get %s secret: %v", managedFusionSecretName, err)
+	}
+	// Check if the structure of managedFusionSecret is valid or not
+	managedFusionSecretData := r.managedFusionSecret.Data
+	managedFusionDeploymentPagerDutyConfig, exist := managedFusionSecretData["pager_duty_config"]
+	if !exist {
+		return ctrl.Result{}, fmt.Errorf("%s secret does not contain pager_duty_config entry", managedFusionSecretName)
+	}
+	if err := yaml.Unmarshal(managedFusionDeploymentPagerDutyConfig, r.pagerDutyConfigData); err != nil {
+		return ctrl.Result{}, fmt.Errorf("%s.pager_duty_config is not in a valid yaml format: %v", managedFusionSecretName, err)
+	}
+	managedFusionDeploymentSMTPConfig, exist := managedFusionSecretData["smtp_config"]
+	if !exist {
+		return ctrl.Result{}, fmt.Errorf("%s secret does not contain smtp_config entry", managedFusionSecretName)
+	}
+	if err := yaml.Unmarshal(managedFusionDeploymentSMTPConfig, r.smtpConfigData); err != nil {
+		return ctrl.Result{}, fmt.Errorf("%s.smtp_config is not in a valid yaml format: %v", managedFusionSecretName, err)
+	}
```

Comment:
Let's try to minimize function scope variable pollution, this will also enable us to use shorter names as they are relevant only in the sub-scope. 

In addition, in go, we usually use the name `found` and not `exists` for map key checks. Most of go online documentation uses it  

 ```suggestion
	if config, found := r.managedFusionSecret.Data["pager_duty_config"]; !found {
		return ctrl.Result{}, fmt.Errorf("%s secret does not contain pager_duty_config entry", managedFusionSecretName)
	} else if err := yaml.Unmarshal(config, r.pagerDutyConfigData); err != nil {
		return ctrl.Result{}, fmt.Errorf("%s.pager_duty_config is not in a valid yaml format: %v", managedFusionSecretName, err)
	}
	
	if config, found := r.managedFusionSecret.Data["smtp_config"]; !found {
		return ctrl.Result{}, fmt.Errorf("%s secret does not contain smtp_config entry", managedFusionSecretName)
	} else if err := yaml.Unmarshal(config, r.smtpConfigData); err != nil {
		return ctrl.Result{}, fmt.Errorf("%s.smtp_config is not in a valid yaml format: %v", managedFusionSecretName, err)
	}
```

Reply:
Updated